libraries/Microsoft.Bot.Builder.Dialogs/DialogContextState.cs

```diff
@@ -75,11 +87,33 @@ public StateMap Dialog
             }
         }
 
-        public DialogContextState(DialogContext dc, StateMap userState, StateMap conversationState)
+        [JsonProperty(PropertyName = "turn")]
+        public Dictionary<string, object> Turn { get; set; }
+
+        public ICollection<string> Keys => new[] { "user", "conversation", "dialog", "turn" };
+
+        public ICollection<object> Values => new[] { User, Conversation, Dialog, Turn };
+
+        public int Count => 3;
+
+        public bool IsReadOnly => true;
+
+        public object this[string key]
         {
-            this.dialogContext = dc ?? throw new ArgumentNullException(nameof(dc));
-            this.User = userState;
-            this.Conversation = conversationState;
+            get
+            {
+                if (TryGetValue(key, out object result))
+                {
+                    return result;
+                }
+
+                return null;
+            }
+
+            set
+            {
```

Comment:
don't we want Set? if you don't I'd suggest to not expose a setter so it results in a compile time error if somebody sets. Otherwise, this will trigger a runtime error.